src/main/java/uk/gov/homeoffice/digital/sas/timecard/validators/timeentry/TimeEntryValidator.java

```diff
@@ -30,6 +35,87 @@ public boolean isValid(Object value, ConstraintValidatorContext context) {
-    return timeEntryClashes.isEmpty();
+    if (!timeEntryClashes.isEmpty()) {
+      ClashingProperty clashingProperty = getClashingProperty(timeEntry, timeEntryClashes);
+      var payload = timeEntryClashes.stream().map(this::transformTimeEntry)
+          .collect(Collectors.toCollection(ArrayList::new));
+
+      HibernateConstraintValidatorContext hibernateContext =
+          context.unwrap(HibernateConstraintValidatorContext.class);
+
+      hibernateContext.disableDefaultConstraintViolation();
+      hibernateContext.withDynamicPayload(payload);
+      hibernateContext
+          .buildConstraintViolationWithTemplate(
+              "Time periods must not overlap with another time period")
+          .addPropertyNode(clashingProperty.toString())
+          .addConstraintViolation();
+      return false;
+    }
+    return true;
+  }
+
+  private JSONObject transformTimeEntry(TimeEntry timeEntry) {
+    var result = new JSONObject();
+    result.put("startTime", timeEntry.getActualStartTime());
+    result.put("endTime", timeEntry.getActualEndTime());
+    result.put("timePeriodTypeId", timeEntry.getTimePeriodTypeId());
+    return result;
+  }
+
+  private ClashingProperty getClashingProperty(
+      TimeEntry timeEntry, List<TimeEntry> timeEntryClashes) {
+    var startTimeClash = false;
+    var endTimeClash = false;
+
+    for (TimeEntry timeEntryClash : timeEntryClashes) {
+      if (timeEntry.getActualStartTime().equals(timeEntryClash.getActualStartTime())) {
+        startTimeClash = true;
+      }
+
+      if (timeEntryClash.getActualEndTime() != null
+          && isStartTimeInTimeEntry(timeEntry.getActualStartTime(), timeEntryClash)) {
+        startTimeClash = true;
+      }
+
+      if (timeEntry.getActualEndTime() != null
+          && timeEntryClash.getActualEndTime() != null
+          && isEndTimeInTimeEntry(timeEntry.getActualEndTime(), timeEntryClash)) {
+        endTimeClash = true;
+      }
+
+      if (timeEntry.getActualEndTime() != null
+          && timeEntryClash.getActualEndTime() != null
+          && isStartTimeInTimeEntry(timeEntryClash.getActualStartTime(), timeEntry)
+      ) {
+        endTimeClash = true;
+      }
+    }
+
+    if (startTimeClash && endTimeClash) {
+      return ClashingProperty.startAndEndTime;
+    }
+    if (startTimeClash) {
+      return ClashingProperty.startTime;
+    }
+    if (endTimeClash) {
+      return ClashingProperty.endTime;
+    }
+
+    return null;
+  }
+
+  private boolean isEndTimeInTimeEntry(Date endTime, TimeEntry timeEntry) {
+    return (endTime.after(timeEntry.getActualStartTime())
+        && endTime.getTime() <= timeEntry.getActualEndTime().getTime());
+  }
+
+  private boolean isStartTimeInTimeEntry(Date startTime, TimeEntry timeEntry) {
+    return (startTime.getTime() >= timeEntry.getActualStartTime().getTime()
+        && startTime.before(timeEntry.getActualEndTime()));
+  }
+
+  enum ClashingProperty {
+    startTime, endTime, startAndEndTime
```

Comment:
These properties should follow the naming convention for constants. 
We can just add a e.g. name property with the actual name (as they are here) and a getter.
Then call the getter on line 51
```suggestion
    START_TIME("startTime"), ...
```

Reply:
I've changed this now :)